src/sankey_diagram_plotly.py

```diff
@@ -1,172 +1,415 @@
+        "file_patterns": [".*Test.java"],
+    },
+    "pytest": {
+        "dependency_patterns": [],
+        "config_files": ["pytest.ini", "tox.ini", "pyproject.toml"],
+        "file_patterns": ["test_.*.py", ".*_test.py"],
+    },
+    "Mocha": {
+        "dependency_patterns": ["mocha"],
+        "config_files": ["package.json", ".mocharc.json", ".mocharc.js"],
+        "file_patterns": ["test/.*.js"],
+    },
+}
+
+
+def detect_test_runners(repo_path):
+    runner_details = {
+        runner: {"dependency_patterns": 0, "config_files": 0, "file_patterns": 0}
+        for runner in test_runners
+    }
 
-from utils.git_utils import get_working_directory_or_git_root
-from utils.initial_data_preparation import (
-    filter_out_incomplete_urls,
-    remove_duplicates,
-)
+    for root, dirs, files in os.walk(repo_path):
+        for runner, indicators in test_runners.items():
+            # Check for configuration files
+            for config_file in indicators["config_files"]:
+                if config_file in files:
+                    runner_details[runner]["config_files"] += 1
 
+            # Check for file patterns
+            for pattern in indicators["file_patterns"]:
+                matching_files = [file for file in files if re.match(pattern, file)]
+                runner_details[runner]["file_patterns"] += len(matching_files)
 
-def process_domain_df(domain_df):
-    # Removing duplicates
-    unique_df, num_duplicates = remove_duplicates(domain_df)
+            # Check for dependencies in the relevant configuration files
+            for dep_pattern in indicators["dependency_patterns"]:
+                for file in files:
+                    if file in indicators["config_files"]:
+                        path = os.path.join(root, file)
+                        if os.path.exists(path):
+                            with open(path, "r") as file_content:
+                                content = file_content.read()
+                                if dep_pattern in content:
+                                    runner_details[runner]["dependency_patterns"] += 1
 
-    # Filtering out incomplete URLs
-    complete_df = filter_out_incomplete_urls(unique_df)
-    num_incomplete_urls = len(unique_df) - len(complete_df)
+    return runner_details
 
-    # Dropping null values
-    no_null_df = complete_df.dropna()
-    num_null_values = len(complete_df) - len(no_null_df)
 
-    return {
-        "duplicates_count": num_duplicates,
-        "incomplete_urls_count": num_incomplete_urls,
-        "null_values_count": num_null_values,
-    }
+def add_node(node_name):
+    global counter
+    if node_name not in node_dict:
+        node_dict[node_name] = counter
+        counter += 1
 
 
 # Setting up the working directory and logger
 working_directory = get_working_directory_or_git_root()
 logger.info(f"Working directory is: {working_directory}")
 
+df = pd.read_csv(working_directory / "data" / "updated_local_github_df_test_count.csv")
```

Comment:
This comment came from reading the code later in this file and is a general comment in terms of the structure and testability of this code.

Currently the code is straight-line rather than packaged as functions. There's a large block of code from roughly line 77 where the columns are going to be initialised until roughly line 394 which does one thing then another, then another... How about wrapping that into a function that accepts a dataframe? then we could create some tests that exercise that function e.g. to pass it pre-determined dataframes to check that the function doesn't fail with an exception e.g. when there are 0 rows that match various criterion. I expect we could then modularise the code and add tests for the modules.

PS: I appreciate this code is still a work-in-progress where we're currently exploring how to create the data to feed Plotly's Sankey diagram so some of the current code may be short-lived and premature for modularisation, etc.

Reply:
These sections have been wrapped in the function: `prepare_sankey_data`